core/src/main/java/com/linecorp/armeria/common/util/AbstractOption.java

```diff
@@ -235,40 +236,65 @@ private static final class Pool {
 
         private final Class<?> type;
         private final BiMap<String, AbstractOption<?, ?, ?>> options;
+        private final ReentrantLock reentrantLock = new ReentrantLock();
 
         Pool(Class<?> type) {
             this.type = type;
             options = HashBiMap.create();
         }
 
-        synchronized <T extends AbstractOption<T, U, V>, U extends AbstractOptionValue<U, T, V>, V>
+        <T extends AbstractOption<T, U, V>, U extends AbstractOptionValue<U, T, V>, V>
         T define(Factory<T, U, V> optionFactory, String name, V defaultValue,
                  Function<V, V> validator, BiFunction<U, U, U> mergeFunction) {
-            final AbstractOption<?, ?, ?> oldOption = options.get(name);
-            if (oldOption != null) {
-                throw new IllegalStateException(
+            lock();
+            try {
+                final AbstractOption<?, ?, ?> oldOption = options.get(name);
+                if (oldOption != null) {
+                    throw new IllegalStateException(
                         '\'' + type.getName() + '#' + name + "' exists already.");
-            }
+                }
 
-            final T newOption = optionFactory.get(name, defaultValue, validator, mergeFunction);
-            checkArgument(type.isInstance(newOption),
-                          "OptionFactory.newOption() must return an instance of %s.", type);
-            options.put(name, newOption);
-            return newOption;
+                final T newOption = optionFactory.get(name, defaultValue, validator, mergeFunction);
+                checkArgument(type.isInstance(newOption),
+                    "OptionFactory.newOption() must return an instance of %s.", type);
+                options.put(name, newOption);
+                return newOption;
+            }
+            finally {
```

Comment:
This violates our checkstyle rule.
```suggestion
            } finally {
```